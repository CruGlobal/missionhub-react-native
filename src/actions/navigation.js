import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
