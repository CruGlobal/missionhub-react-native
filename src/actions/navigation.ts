/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
