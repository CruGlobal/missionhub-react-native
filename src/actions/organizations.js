import {
  GET_ORGANIZATIONS_CONTACTS_REPORT,
  GET_ORGANIZATION_MEMBERS,
  DEFAULT_PAGE_LIMIT,
  LOAD_ORGANIZATIONS,
  ORGANIZATION_CONTACTS_SEARCH,
  SURVEY_CONTACTS_SEARCH,
  RESET_ORGANIZATION_CONTACTS,
  RESET_SURVEY_CONTACTS,
} from '../constants';
import { timeFilter } from '../utils/filters';
import { organizationSelector } from '../selectors/organizations';

import callApi, { REQUESTS } from './api';

const getOrganizationsQuery = {
  limit: 100,
  include: '',
  filters: {
    descendants: false,
  },
};

export function getMyCommunities() {
  return async dispatch => {
    await dispatch(getMyOrganizations());
    return dispatch(getOrganizationsContactReports());
  };
}

export function getMyOrganizations() {
  return async (dispatch, getState) => {
    const { response: orgs } = await dispatch(
      callApi(REQUESTS.GET_ORGANIZATIONS, getOrganizationsQuery),
    );
    const orgOrder = getState().auth.person.user.organization_order;

    if (orgOrder) {
      orgs.sort((a, b) => {
        const aIndex = orgOrder.indexOf(a.id);
        const bIndex = orgOrder.indexOf(b.id);

        if (aIndex === -1) {
          return bIndex === -1 ? 0 : 1;
        }

        if (bIndex === -1) {
          return aIndex === -1 ? 0 : -1;
        }

        return aIndex > bIndex ? 1 : -1;
      });
    }

    dispatch({
      type: LOAD_ORGANIZATIONS,
      orgs,
    });
  };
}

export function getOrganizationsContactReports() {
  return async dispatch => {
    const { response } = await dispatch(
      callApi(REQUESTS.GET_ORGANIZATION_INTERACTIONS_REPORT, { period: 'P1W' }),
    );

    dispatch({
      type: GET_ORGANIZATIONS_CONTACTS_REPORT,
      reports: response.map(r => ({
        id: `${r.organization_id}`,
        contactsCount: r.contact_count,
        unassignedCount: r.unassigned_count,
        uncontactedCount: r.uncontacted_count,
      })),
    });
    return response;
  };
}

export function getOrganizationContacts(orgId, name, filters = {}) {
  const query = {
    filters: {
      permissions: 'no_permission',
      organization_ids: orgId,
    },
    include:
      'reverse_contact_assignments,reverse_contact_assignments.organization,organizational_permissions',
  };
  if (name) {
    query.filters.name = name;
  }

  const answerFilters = getAnswersFromFilters(filters);
  if (answerFilters) {
    query.filters.answer_sheets = { answers: answerFilters };
  }
  if (filters.survey) {
    query.filters.answer_sheets = {
      ...(query.filters.answer_sheets || {}),
      survey_ids: filters.survey.id,
    };

    // If there is a survey AND we're filtering by time, apply the time filter to the answer_sheets
    if (filters.time) {
      const dates = timeFilter(filters.time.value);
      query.filters.answer_sheets = {
        ...(query.filters.answer_sheets || {}),
        created_at: [dates.first, dates.last],
      };
    }
  } else {
    // TODO: Enable this when the API supports sorting contacts by `updated_at`
    //   if (filters.time) {
    //     const dates = timeFilter(filters.time.value);
    //     query.filters.updated_at = [dates.first, dates.last];
    //   }
  }
  if (filters.gender) {
    query.filters.genders = filters.gender.id;
  }
  if (filters.archived) {
    query.filters.include_archived = true;
  }
  if (filters.unassigned) {
    query.filters.assigned_tos = 'unassigned';
  }
  if (filters.uncontacted) {
    query.filters.statuses = 'uncontacted';
  }
  if (filters.labels) {
    query.filters.label_ids = filters.labels.id;
  }
  if (filters.groups) {
    query.filters.group_ids = filters.groups.id;
  }

  return async (dispatch, getState) => {
    const { organizations } = getState();
    const org = organizationSelector({ organizations }, { orgId });
    const pagination = org.contactPagination || { page: 0, hasNextPage: true };

    const offset = DEFAULT_PAGE_LIMIT * pagination.page;

    query.page = {
      limit: DEFAULT_PAGE_LIMIT,
      offset,
    };

    const { response, meta } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_LIST, query),
    );

    return dispatch({
      type: filters.survey
        ? SURVEY_CONTACTS_SEARCH
        : ORGANIZATION_CONTACTS_SEARCH,
      orgId,
      contacts: response,
      query,
      meta,
    });
  };
}

export function reloadOrganizationContacts(orgId, name, filters = {}) {
  return (dispatch, getState) => {
    const { organizations } = getState();
    const org = organizationSelector({ organizations }, { orgId });

    if (filters.survey && filters.survey.id) {
      const surveyId = filters.survey.id;
      if (
        org.surveys &&
        org.surveys[surveyId] &&
        org.surveys[surveyId].contactPagination
      ) {
        dispatch(resetContactPagination(orgId, surveyId));
      }
    }

    if (org && org.contactPagination) {
      dispatch(resetContactPagination(orgId));
    }
    return dispatch(getOrganizationContacts(orgId, name, filters));
  };
}

const resetContactPagination = (orgId, surveyId = undefined) => {
  return {
    type: surveyId ? RESET_SURVEY_CONTACTS : RESET_ORGANIZATION_CONTACTS,
    orgId,
    surveyId,
  };
};

//each question/answer filter must be in the URL in the form:
//filters[answers][questionId][]=answerTexts
function getAnswersFromFilters(filters) {
  const arrFilters = Object.keys(filters).map(k => filters[k]);
  const answers = arrFilters.filter(f => f.isAnswer);
  if (answers.length === 0) {
    return null;
  }
  let answerFilters = {};
  answers.forEach(f => {
    answerFilters[f.id] = [f.text];
  });
  return answerFilters;
}

//todo probably should start storing this stuff in Redux
export function getOrganizationMembers(orgId, query = {}) {
  const newQuery = {
    ...query,
    filters: {
      permissions: 'admin,user',
      organization_ids: orgId,
    },
    include: 'organizational_permissions',
  };
  return async dispatch => {
    const { response: members, meta } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_LIST, newQuery),
    );

    const memberIds = members.map(m => m.id);
    const reportQuery = {
      people_ids: memberIds.join(','),
      period: 'P1Y',
      organization_ids: orgId,
    };
    const { response: reportResponse } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_INTERACTIONS_REPORT, reportQuery),
    );

    // Get an object with { [key = person_id]: [value = { counts }] }
    const reportsCountObj = reportResponse.reduce(
      (p, n) => ({
        ...p,
        [`${n.person_id}`]: {
          contact_count: n.contact_count,
          uncontacted_count: n.uncontacted_count,
          contacts_with_interaction_count: n.contacts_with_interaction_count,
        },
      }),
      {},
    );
    // Merge the counts into the members array
    const membersWithCounts = members.map(m => ({
      ...m,
      ...(reportsCountObj[m.id] || {}),
    }));

    dispatch({
      type: GET_ORGANIZATION_MEMBERS,
      orgId,
      members: membersWithCounts,
      query: newQuery,
      meta,
    });
    return membersWithCounts;
  };
}

export function getOrganizationMembersNextPage(orgId) {
  return (dispatch, getState) => {
    const { page, hasNextPage } = getState().organizations.membersPagination;
    if (!hasNextPage) {
      // Does not have more data
      return Promise.resolve();
    }
    const query = {
      page: {
        limit: DEFAULT_PAGE_LIMIT,
        offset: DEFAULT_PAGE_LIMIT * page,
      },
    };
    return dispatch(getOrganizationMembers(orgId, query));
  };
}

export function addNewContact(data) {
  return (dispatch, getState) => {
    const {
      person: { id: myId },
    } = getState().auth;
    if (!data || !data.firstName) {
      return Promise.reject(
        `Invalid Data from addNewContact: no data or no firstName passed in`,
      );
    }
    let included = [];
    included.push({
      type: 'contact_assignment',
      attributes: {
        assigned_to_id: myId,
        organization_id: data.orgId || undefined,
      },
    });
    if (data.orgId) {
      included.push({
        type: 'organizational_permission',
        attributes: {
          organization_id: data.orgId,
          permission_id: data.orgPermission && data.orgPermission.permission_id,
        },
      });
    }
    if (data.email) {
      included.push({
        type: 'email',
        attributes: { email: data.email },
      });
    }
    if (data.phone) {
      included.push({
        type: 'phone_number',
        attributes: {
          number: data.phone,
          location: 'mobile',
        },
      });
    }
    const bodyData = {
      data: {
        type: 'person',
        attributes: {
          first_name: data.firstName,
          last_name: data.lastName || undefined,
          gender: data.gender || undefined,
        },
      },
      included,
    };
    const query = {};
    return dispatch(callApi(REQUESTS.ADD_NEW_PERSON, query, bodyData));
  };
}
