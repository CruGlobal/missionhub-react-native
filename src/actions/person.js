import callApi, { REQUESTS } from './api';
import { PERSON_FIRST_NAME_CHANGED, PERSON_LAST_NAME_CHANGED } from '../constants';


export function personFirstNameChanged(firstName) {
  return {
    type: PERSON_FIRST_NAME_CHANGED,
    personFirstName: firstName,
  };
}

export function personLastNameChanged(lastName) {
  return {
    type: PERSON_LAST_NAME_CHANGED,
    personLastName: lastName,
  };
}

export function saveNotes(personId, notes, firstSave) {
  return (dispatch, getState) => {
    const myId = getState().auth.personId;
    if (!personId || !notes) {
      return Promise.reject('InvalidData', personId, notes);
    }

    const bodyData = {
      data: {
        type: 'person_note',
        attributes: {
          content: notes,
        },
      },
      relationships: {
        person: {
          data: {
            type: 'person',
            id: personId,
          },
        },
      },
      user: {
        data: {
          type: 'user',
          id: myId,
        },
      },
    };

    const query = {};

    if (firstSave) return dispatch(callApi(REQUESTS.ADD_PERSON_NOTES, query, bodyData));
    return dispatch(callApi(REQUESTS.UPDATE_PERSON_NOTES, query, bodyData));
  };
}