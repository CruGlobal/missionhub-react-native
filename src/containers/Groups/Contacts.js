import React, { Component } from 'react';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import PropTypes from 'prop-types';

import { navigatePush } from '../../actions/navigation';
import {
  getOrganizationContacts,
  reloadOrganizationContacts,
} from '../../actions/organizations';
import { navToPersonScreen } from '../../actions/person';
import { Flex } from '../../components/common';
import SearchList from '../../components/SearchList';
import ContactItem from '../../components/ContactItem';
import { searchRemoveFilter } from '../../utils/filters';
import { organizationSelector } from '../../selectors/organizations';
import { contactsInOrgSelector } from '../../selectors/people';

import { SEARCH_CONTACTS_FILTER_SCREEN } from './ContactsFilter';

@translate('groupsContacts')
class Contacts extends Component {
  constructor(props) {
    super(props);
    const { t } = props;

    this.state = {
      filters: {
        // Default filters
        unassigned: {
          id: 'unassigned',
          selected: true,
          text: t('searchFilter:unassigned'),
        },
        // TODO: temporarily remove this until the API supports it
        // time: { id: 'time30', text: t('searchFilter:time30') },
      },
      defaultResults: [],
    };
  }

  componentDidMount() {
    // TODO: Only do this when this tab is focused to improve performance
    // Use the default filters to load in these people
    this.loadContactsWithFilters();
  }

  loadContactsWithFilters = async () => {
    const contacts = await this.handleSearch('');

    this.setState({ defaultResults: contacts });
  };

  handleRemoveFilter = key => {
    return searchRemoveFilter(this, key, ['unassigned', 'time']);
  };

  handleFilterPress = () => {
    const { dispatch, organization } = this.props;
    const { filters } = this.state;
    dispatch(
      navigatePush(SEARCH_CONTACTS_FILTER_SCREEN, {
        onFilter: this.handleChangeFilter,
        organization,
        filters,
      }),
    );
  };

  handleChangeFilter = filters => {
    this.setState({ filters }, () => {
      // Run the search every time a filter option changes
      if (this.searchList && this.searchList.getWrappedInstance) {
        this.searchList.getWrappedInstance().search();
      }
    });
  };

  handleSearch = async text => {
    const { dispatch, organization } = this.props;
    const { filters } = this.state;
    return await dispatch(
      reloadOrganizationContacts(organization.id, text, filters),
    );
  };

  handleLoadMore = async text => {
    const { dispatch, organization } = this.props;
    const { filters } = this.state;
    return await dispatch(
      getOrganizationContacts(organization.id, text, filters),
    );
  };

  handleSelect = person => {
    const { dispatch, organization } = this.props;
    dispatch(navToPersonScreen(person, organization));
  };

  listRef = c => (this.searchList = c);

  renderItem = ({ item }) => (
    <ContactItem
      organization={this.props.organization}
      contact={item}
      onSelect={this.handleSelect}
    />
  );

  render() {
    const { t, contactList } = this.props;
    const { filters, defaultResults } = this.state;
    return (
      <Flex value={1}>
        <SearchList
          ref={this.listRef}
          defaultData={defaultResults}
          onFilterPress={this.handleFilterPress}
          listProps={{
            renderItem: this.renderItem,
          }}
          onSearch={this.handleSearch}
          onRemoveFilter={this.handleRemoveFilter}
          onLoadMore={this.handleLoadMore}
          filters={filters}
          placeholder={t('searchPlaceholder')}
          results={contactList}
        />
      </Flex>
    );
  }
}

Contacts.propTypes = {
  organization: PropTypes.object.isRequired,
  contactList: PropTypes.array.isRequired,
  pagination: PropTypes.object.isRequired,
};

const mapStateToProps = ({ organizations, contacts }, { orgId }) => {
  const organization = organizationSelector({ organizations }, { orgId });
  const contactList = contactsInOrgSelector({ contacts }, { organization });
  const pagination = organization.contactPagination || {
    page: 0,
    hasNextPage: true,
  };

  return {
    organization,
    contactList,
    pagination,
  };
};

export default connect(mapStateToProps)(Contacts);
