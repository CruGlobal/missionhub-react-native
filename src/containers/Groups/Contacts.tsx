import React, { Component } from 'react';
import { View } from 'react-native';
import { connect } from 'react-redux-legacy';
import { withTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

import { navigatePush } from '../../actions/navigation';
import { getOrganizationContacts } from '../../actions/organizations';
import { navToPersonScreen } from '../../actions/person';
import SearchList from '../../components/SearchList';
import PersonListItem from '../../components/PersonListItem';
import { organizationSelector } from '../../selectors/organizations';
import { searchRemoveFilter, unassignedFilter } from '../../utils/filters';
import { buildUpdatedPagination } from '../../utils/pagination';

import { SEARCH_CONTACTS_FILTER_SCREEN } from './ContactsFilter';
import styles from './styles';

// @ts-ignore
@withTranslation('groupsContacts')
class Contacts extends Component {
  state = {
    pagination: {
      page: 0,
      hasMore: true,
    },
    filters: {
      // Default filters
      // @ts-ignore
      unassigned: unassignedFilter(this.props.t, true),
      // TODO: temporarily remove this until the API supports it
      // time: thirtyDaysFilter(t),
    },
    defaultResults: [],
  };

  componentDidMount() {
    // TODO: Only do this when this tab is focused to improve performance
    // Use the default filters to load in these people
    this.loadContactsWithFilters();
  }

  loadContactsWithFilters = async () => {
    const contacts = await this.handleLoadMore('');
    this.setState({ defaultResults: contacts });
  };

  // @ts-ignore
  handleRemoveFilter = key => {
    // @ts-ignore
    return searchRemoveFilter(this, key, ['unassigned', 'time']);
  };

  handleFilterPress = () => {
    // @ts-ignore
    const { dispatch, organization } = this.props;
    const { filters } = this.state;
    dispatch(
      navigatePush(SEARCH_CONTACTS_FILTER_SCREEN, {
        onFilter: this.handleChangeFilter,
        organization,
        filters,
      }),
    );
  };

  // @ts-ignore
  handleChangeFilter = filters => {
    this.setState({ filters });
    this.handleRefreshSearchList();
  };

  // @ts-ignore
  handleSearch = async text => {
    await this.setState({ pagination: { page: 0, hasMore: true } });
    return this.handleLoadMore(text);
  };

  handleRefreshSearchList = () =>
    // @ts-ignore
    this.searchListSearch && this.searchListSearch();

  // @ts-ignore
  handleLoadMore = async text => {
    // @ts-ignore
    const { dispatch, organization } = this.props;
    const { filters, pagination } = this.state;

    const result = await dispatch(
      // @ts-ignore
      getOrganizationContacts(organization.id, text, pagination, filters),
    );

    const { meta, response } = result;

    this.setState({ pagination: buildUpdatedPagination(meta, pagination) });

    return response;
  };

  // @ts-ignore
  handleSelect = person => {
    // @ts-ignore
    const { dispatch, organization } = this.props;
    dispatch(
      navToPersonScreen(person, organization, {
        onAssign: this.handleRefreshSearchList,
      }),
    );
  };

  // @ts-ignore
  setSearch = search => (this.searchListSearch = search);

  // @ts-ignore
  renderItem = ({ item }) => (
    <PersonListItem
      // @ts-ignore
      organization={this.props.organization}
      person={item}
      onSelect={this.handleSelect}
    />
  );

  render() {
    // @ts-ignore
    const { t } = this.props;
    const { filters, defaultResults } = this.state;
    return (
      <View style={styles.pageContainer}>
        <SearchList
          // @ts-ignore
          setSearch={this.setSearch}
          defaultData={defaultResults}
          onFilterPress={this.handleFilterPress}
          listProps={{
            renderItem: this.renderItem,
          }}
          onSearch={this.handleSearch}
          onRemoveFilter={this.handleRemoveFilter}
          onLoadMore={this.handleLoadMore}
          filters={filters}
          placeholder={t('searchPlaceholder')}
        />
      </View>
    );
  }
}

// @ts-ignore
Contacts.propTypes = {
  organization: PropTypes.object.isRequired,
};

// @ts-ignore
const mapStateToProps = ({ organizations }, { orgId }) => ({
  organization: organizationSelector({ organizations }, { orgId }),
});

export default connect(mapStateToProps)(Contacts);
