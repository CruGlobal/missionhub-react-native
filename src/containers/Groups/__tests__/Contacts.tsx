import React from 'react';

import Contacts from '../Contacts';
import {
  createThunkStore,
  renderShallow,
  testSnapshotShallow,
} from '../../../../testUtils';
import { navigatePush } from '../../../actions/navigation';
import { navToPersonScreen } from '../../../actions/person';
import { getOrganizationContacts } from '../../../actions/organizations';
import { buildUpdatedPagination } from '../../../utils/pagination';

jest.mock('../../../actions/navigation', () => ({
  navigatePush: jest.fn(() => ({ type: 'test' })),
}));
jest.mock('../../../actions/person', () => ({
  navToPersonScreen: jest.fn(() => ({ type: 'test' })),
}));

jest.mock('../../../actions/organizations');
jest.mock('../../../utils/pagination');

const people = [{ id: '1' }, { id: '2' }];

const orgId = '1';
const organization = { id: orgId, name: 'Test Org' };

const store = createThunkStore({ organizations: { all: [organization] } });

// @ts-ignore
getOrganizationContacts.mockReturnValue({
  type: 'got org contacts',
  response: people,
});
// @ts-ignore
buildUpdatedPagination.mockReturnValue({});

describe('Contacts', () => {
  const component = <Contacts orgId={orgId} />;

  it('should render correctly', () => {
    testSnapshotShallow(component, store);
  });

  it('should load contacts with filter', async () => {
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    instance.handleSearch = jest.fn(() => Promise.resolve(people));
    // @ts-ignore
    await instance.loadContactsWithFilters();
    // @ts-ignore
    expect(instance.state.defaultResults).toEqual(people);
  });

  it('should handleFilterPress correctly', () => {
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    instance.handleFilterPress();
    expect(navigatePush).toHaveBeenCalled();
  });

  it('should handleChangeFilter correctly', () => {
    const instance = renderShallow(component, store).instance();
    const newFilters = {
      filter1: { id: '1' },
    };
    // @ts-ignore
    instance.handleChangeFilter(newFilters);
    // @ts-ignore
    expect(instance.state.filters).toEqual(newFilters);
  });

  it('should handleSearch correctly', async () => {
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    const result = await instance.handleSearch('test');
    expect(result).toEqual(people);
  });

  it('should handleRemoveFilter correctly', async () => {
    const instance = renderShallow(component, store).instance();
    instance.setState({
      filters: {
        filter1: { id: '1', text: 'Last 30 days' },
        filter2: { id: '2', text: 'Last 7 days' },
      },
    });
    // @ts-ignore
    await instance.handleRemoveFilter('filter1');

    // @ts-ignore
    expect(Object.keys(instance.state.filters)).toHaveLength(1);
    // @ts-ignore
    expect(instance.state.filters.filter2).toMatchObject({
      id: '2',
      text: 'Last 7 days',
    });
  });

  it('should handleSelect correctly', () => {
    const person = people[0];
    const screen = renderShallow(component, store);
    const listItem = screen
      .childAt(1)
      .props()
      .listProps.renderItem({ item: person });

    listItem.props.onSelect(person);

    expect(navToPersonScreen).toHaveBeenCalledWith(person, organization, {
      // @ts-ignore
      onAssign: screen.instance().handleRefreshSearchList,
    });
  });
  it('should set searchListSearch when setSearch is called', () => {
    const instance = renderShallow(component, store).instance();
    const childSearch = () => {};
    // @ts-ignore
    instance.setSearch(childSearch);
    // @ts-ignore
    expect(instance.searchListSearch).toEqual(childSearch);
  });
  it('should render item', () => {
    const instance = renderShallow(component, store).instance();
    // @ts-ignore
    const renderedItem = instance.renderItem({ item: people[0] });
    expect(renderedItem).toMatchSnapshot();
  });
});
