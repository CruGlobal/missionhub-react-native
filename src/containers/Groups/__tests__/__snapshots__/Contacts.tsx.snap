// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <InfoModal
    action={[Function]}
    buttonLabel="FIND THEM HERE"
    title="We are moving Contacts to the web!"
    titleStyle={
      Object {
        "maxWidth": 200,
      }
    }
  />
  <withI18nextTranslation(SearchList)
    defaultData={Array []}
    filters={
      Object {
        "unassigned": Object {
          "id": "unassigned",
          "selected": true,
          "text": "Unassigned",
        },
      }
    }
    listProps={
      Object {
        "renderItem": [Function],
      }
    }
    onFilterPress={[Function]}
    onLoadMore={[Function]}
    onRemoveFilter={[Function]}
    onSearch={[Function]}
    placeholder="Search Contacts"
    setSearch={[Function]}
  />
</View>
`;

exports[`Contacts should render item 1`] = `
<PersonListItem
  onSelect={[Function]}
  organization={
    Object {
      "id": "1",
      "name": "Test Org",
    }
  }
  person={
    Object {
      "id": "1",
    }
  }
/>
`;
